loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+- Each plugin is given a chance to add any plugin candidates
+    - Each new plugin added this way is given a chance to add any plugin candidates.
+      Rinse and repeat.
+- Each plugin is given a chance to add any required or tentative mod candidates,
+  rules, or options
+- Quilt Loader will evaluate the sat4j tree
+    - Each plugin will be given a chance to claim unsatisfied rules for resolution
+    - Each plugins resolvers will be run sequentially
+
+If any changes to the mod set have taken place during a given cycle, another cycle
+will be run. Otherwise, the game will either launch or crash, depending on whether
+or not all rules have been satisfied.
+
+
+## API Surface
+
+Below is a tentative plan of what the surface level API classes might look like.
+Everything here is subject (and likely) to change, but helps provide a visual
+idea of how developers might use this feature down the road.
+
+<details>
+    <summary>QuiltPluginContext</summary>
+
+```java
+/**
+* Passed to loader plugins to define what actions they are able to take.
+*/
+public sealed interface QuiltPluginContext permits QuiltPluginContextImpl {
+/**
+* The plugin that this context is for. This method is useless, it just indicates that every other method here has an 
+* implicit paramater of "The Loader Plugin" for the UI / logging to use in some way
+*/
+QuiltLoaderPlugin plugin();
+
+void addCandidate(ModCandidate candidate);
+
+void addCandidate(PluginCandidate candidate);
+
+/**
+* Adds a tentative mod candidate which indicates that downloading / fetching a new mod will fix a rule somewhere.
+* This tentative mod won't be kept around to the next cycle - instead the resolver is called to actually download
+* the mod if {@link QuiltLoaderPlugin#canResolve} returns true after each plugin has been checked.
+*/
+void addTentativeCandidate(String group, String modId, Version version, CandidateResolver resolver);
+
+/**
+* Adds a rule to the current solver.
+*/
+void addRule(Rule rule);
+
+/**
+* Adds a LoadOption to the current solver. All existing rules will have Rule#onLoadOptionAdded called, and all plugins 
+* will have ILoaderPlugin#onLoadOptionAdded called.
+*
+* If this is an ITentativeLoadOption then it will be removed at the end of the cycle, and handled by whatever plugin 
+* added it.
+*/
+void addOption(LoadOption option);
+
+/**
+* Gets the metadata for a given mod.
+*
+* Because the mods in QuiltLoader only reference fully loaded mods, this method can be used during the mod loading process
+* to get the metadata for any candidates (tentative or not) that were present prior to this cycle.
+*/
+ModMetadata getMetadata(String modId);
+}
+```
+</details>
+
+<details>
+    <summary>CandidateResolver</summary>
+    
+```java
+interface CandidateResolver {
+/**
+* Attempts to resolve a tentative candidate.
+*
+* @return an error message if resolution fails, null otherwise
+*/ 
+@Nullable String resolve();
```

Comment:
What would resolution do in this context? My interpretation here is things such as downloading mods or remapping a jar, such as with patchwork.

If it's the latter, would it make more sense to return a future?

Reply:
And with a future, a failed future could be returned given resolution failing.